clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+  // identify which pointers are USM, and which are definitely not (and will
+  // crash). For those that will crash, we emit an error. For those that are
+  // unknown, we _could_ emit a gentle note suggesting the user double check
+  // that they are using USM.  But at this time we do not. For safe USM
+  // pointers, we do nothing.
+
+  enum ExprAllocation { Unknown, USM, WillCrash };
+  ExprAllocation howAllocated = Unknown;
+
+  // We will use this to check declarations and assignments.
+  auto VetteCallExpr = [&howAllocated](const Expr *E) {
+    E = E->IgnoreCasts();
+    const CallExpr *CE = dyn_cast<CallExpr>(E);
+    bool updated = false;
+    if (CE) {
+      const FunctionDecl *func = CE->getDirectCallee();
+      auto FullName = func->getQualifiedNameAsString();
+      // Check to see if this function call is one of the USM allocators.
+      if ((FullName.rfind("cl::sycl::malloc", 0) == 0) ||
+          (FullName.rfind("cl::sycl::aligned_alloc", 0) == 0)) {
+        howAllocated = USM;
+        updated = true;
+      } else if (howAllocated == Unknown &&
+                 ((FullName.compare("malloc") == 0) ||
+                  (FullName.compare("calloc") == 0))) {
+        howAllocated = WillCrash;
+        updated = true;
+      }
+    }
+    return updated;
+  };
+  std::function<bool(const Expr *)> AssignF = VetteCallExpr;
+
+  SourceLocation DecLoc = SourceLocation();
+
+  // Try to determine provenance of this Var.
+  const Expr *Init = Var->getAnyInitializer();
+  if (Init) {
+    Init = Init->IgnoreCasts();
+    DecLoc = Init->getExprLoc();
+
+    const CallExpr *CE = dyn_cast<CallExpr>(Init);
+    if (CE) {
+      VetteCallExpr(CE); // Modifies howAllocated via side-effect.
+    } else {
+      // Var has initialization, but not as return result of a function,
+      // disqualify any other obvious bad initialization expressions.
+      const StringLiteral *SL = dyn_cast<StringLiteral>(Init);
+      if (SL)
+        howAllocated = WillCrash;
+
+      const UnaryOperator *UO = dyn_cast<UnaryOperator>(Init);
+      if (UO && (UO->getOpcode() == UO_AddrOf))
+        howAllocated = WillCrash;
+    }
+  }
+  // else: Var does not have local initialization, might be parameter, etc.
+
+  // We may have identified some initialization as unsafe. But that variable
+  // is subject to change. So we look through the statements to see if that
+  // variable is otherwise referenced before capture, or if reassigned it is
+  // assessed with AssignF.
+  DeclContext *DC = Var->getParentFunctionOrMethod();
+  const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(DC);
+  if (FD && FD->hasBody()) {
+    Stmt *TopStmt = FD->getBody(FD);
+    bool match = isVarInStmt(Var, AssignF, TopStmt);
+    if (match)
+      howAllocated = Unknown;
+  } else {
+    howAllocated = Unknown;
+  }
+
+  // Emit diagnostics.
+  if (howAllocated == WillCrash) {
+    Diag(CaptureLoc, diag::err_sycl_illegal_memory_reference);
```

Comment:
I thought you were switching these to delayed diagnostics?  This using them.

Reply:
It's now called _by_ the delayed diagnostic system. So we emit a normal one.

Reply:
That is inverted, thats not how this should work.

Reply:
I ended up reverting and rebasing. Am emitting deferred diagnostics again. But see my standalone comment about this